spec.md

```diff
@@ -0,0 +1,276 @@
+
+The header `X-IPFS-Path` is set to the URL path. `IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path.
+
+### HEAD `/ipfs/<CID>[/<path>]`
+
+The same as GET, but with no response body.
+
+### POST `/ipfs/`
+
+Upload file(s) to IPFS.
+
+#### Headers
+- `X-IPFS-Pin` pin the uploaded file(s) if this header is present
+
+#### Response
+The response body is uploaded, and a 201 Created redirect to `ipfs://<CID>` is returned. The response header `IPFS-Hash` is set to the CID of the uploaded content.
+
+Form data is supported, in which case multiple files could be uploaded, and the returned CID would point to a directory containing those files.
+
+### PUT `/ipfs/<CID>[/<path>]`
+
+Upload a file to an existing path, such as `/ipfs/<CID>/path/to/file.ext`. Form data is supported, in which case the path must point to a directory, such as `/ipfs/<CID>/path/to/dir`.
+
+This can also be used to wrap a file in a directory, by using the empty directory CID: `/ipfs/bafybeiczsscdsbs7ffqz55asqdf3smv6klcw3gofszvwlyarci47bgf354`.
+
+#### Headers
+- `X-IPFS-Pin` pin the resulting new CID if this header is present
+
+#### Response
+A 201 Created redirect to `ipfs://<new root CID>[/<path>]` is returned. The response header `IPFS-Hash` is set to the new root CID that has resulted from the
+file(s) upload.
+
+### DELETE `/ipfs/<CID>/<path>`
+
+Remove the file or directory at the provided path.
+
+#### Headers
+- `X-IPFS-Pin` pin the new CID that results from deletion if this header is present
+
+#### Response
+A 201 Created redirect to `ipfs://<new root CID>[/<path>]` is returned. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipfs/<CID>/path/to/file.ext`, the redirect will be to `ipfs://<new CID>/path/to`.
+
+The response header `IPFS-Hash` is set to the new root CID that has resulted from the deletion.
+
+### DELETE `/ipfs/<CID>`
+
+Unpin the provided CID.
+
+#### Response
+200 OK if successful.
+
+
+## IPNS APIs
+
+
+### GET or HEAD `/ipns/<key/domain>[/<path>]`
+
+#### Options
+Same as GET `/ipfs/<CID>[/<path>]`
+
+#### Response
+Resolves the IPNS key to an IPFS path, and then returns the same response as `/ipfs/`. Also supports resolving custom domains with [DNSLink](https://dnslink.io/).
+
+The header `X-IPFS-Path` is set to the URL path. `IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path. This allows for getting the CID of an IPNS path/key.
+
+### POST `/ipns/<key>[/<path>]`
+
+Update all or part of the content behind an IPNS name. The new content is pinned, and the previous content is unpinned if it exists.
+
+#### Request Body
+A valid IPFS path in the format `/ipfs/<CID>[/<path>]` A naked CID is also valid.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The header `X-IPNS-Path` is set to the `/ipns/<key>[/<path>]`.
+
+### POST `/ipns/localhost[/<path>]`
```

Comment:
When we were talking about this initially, I was imagining that the `localhost` stuff would be for creating keys and resolving them, not so much acting as a proxy for the other methods.

Right now it acts as doing a seamless POST/DELETE/whatever to the resolved key, but I think it'd be useful to do the following:

```
GET /ipns/localhost?key=KEY => Resolve to `ipns://publickey` (or 301 redirect?) if it's created, 404 if it hasn't been
POST /ipns/localhost?key=KEY => Create a new key and resolve to `ipns://publickey`, no-op if it exists
DELETE /ipns/localhost?key=KEY => Delete the key and unpin whatever data it had.
```

I also had a shower thought earlier today about providing a default UI for creating keys if you do a `GET ipns://localhost` without a key and have `text/html` in the `Accept` header. Something like a basic form where you can enter a key name. Probably something I can look into in the future though.

I think this could also give us a path towards importing keys using `POST` down the line.

What do you think of that? Also cc @lidel

Reply:
I'm not sure I really see the point. It seems like the only feature it really offers is key deletion, which I don't think is that important since keys are cheap. And I think it's a pretty nice API for an application to be able to transparently do operations with a key's name instead of having to use the actual key all the time.

If you think this definitely makes more sense though I'm good to change it, I don't think it makes a huge difference either way.

Reply:
Lemme sleep on it. 🤔 I think my gut feeling is that having the key management be it's own thing could be useful.

Particularly for importing and exporting keys down the line.

Reply:
Added to code in f24dfd7 and spec in 80c5972.